### ISSUE TITLE
'Confirm signup' email {{ .Token }} always errors: 'Token has expired or is invalid'

### ISSUE THREAD

**Issue by petedavisdev** [supabase/supabase#12868]
# Bug report

## Describe the bug

I have the same bug in two apps which is causing an error for new users.

I am using {{ .Token }} in both the 'Confirm signup' and 'Magic Link' email templates.
My form uses`signInWithOtp()` and `verifyOtp()`

For a new user's first attempt, they get the 'Confirm signup' email with a 6-digit token as expected, but this first token always results in '401: Token has expired or is invalid'

When they try again, they get the 'Magic Link' email with the 6-digit token, which works.

## To Reproduce

Here is a link to the code in my repo: 

[petedavisdev/fithacker/blob/master/src/pages/Login.vue](https://github.com/petedavisdev/fithacker/blob/master/src/pages/Login.vue)

1. Go to [fithacker.netlify.app/login](https://fithacker.netlify.app/login) or [choosier.app/account](https://choosier.app/account)
2. Enter new email address
3. Copy token from 'Confirm signup' email into form
5. Try again with the same email
6. Copy token from 'Magic number' email into form
7. Logged in successfully

## Expected behavior

{{ .Token }} in 'Confirm signup' email should be valid.

## Screenshots

For a new user's first attempt, they get the 'Confirm signup' email with a 6-digit token as expected, but this first token always results in '401: Token has expired or is invalid'

![image](https://user-images.githubusercontent.com/7900180/223513571-40349cad-5499-48b5-8767-4ccf58734b75.png)

![image](https://user-images.githubusercontent.com/7900180/223513627-48ca0b16-0be8-40b6-bae7-cdd293f75725.png)

## System information

- OS: macOS and Windows
- Version of supabase-js: ^2.10.0
- Version of Node.js: 18

**Comment by GaryAustin1** [comment]
This seems to be a duplicate of this:
https://github.com/supabase/gotrue/issues/905

**Comment by Lewitje** [comment]
I'm getting the same issue, I have 1 form for login and sign up, I think the issue is in the documentation.
The docs say you can pass the argument 'shouldCreateUser', this implies that if I'm trying to sign up or log in it doesn't matter which Type I pass to the verify.

If you pass the type 'signup' and provide the magic link token it doesn't work.
If you pass the type 'magiclink' and provide the sign up token it doesn't work.

**Comment by Lewitje** [comment]
> This seems to be a duplicate of this: [supabase/gotrue#905](https://github.com/supabase/gotrue/issues/905)

Posted within seconds of each other, the linked issue is exactly what I think the solution should be.

**Comment by petedavisdev** [comment]
To clarify, I am using:
`supabase.auth.verifyOtp({ email, token, type: 'magiclink' });`

Looking at [supabase/gotrue/issues/905](https://github.com/supabase/gotrue/issues/905) it seems that will not work for new signup tokens. 

To save having two separate forms, I could see if the above returns an error then try this:
`supabase.auth.verifyOtp({ email, token, type: 'signup' });`

**Comment by GaryAustin1** [comment]
@petedavisdev  Yes that is the issue.  I don't believe you can do the trial test because once the token is used it is invalidated and can't be used again.  (Not positive if it errors it burns the token, but highly likely)

**Comment by petedavisdev** [comment]
@GaryAustin1 it turns out you can try both and it does not burn the token. The two verifications must be completely independent.

![image](https://user-images.githubusercontent.com/7900180/223556429-ea035834-4250-4041-a076-c1fa339152ac.png)

```
const logInResponse = await supabase.auth.verifyOtp({
	email: email.value,
	token: token.value,
	type: 'magiclink',
});

if (logInResponse.error) {
	const signUpResponse = await supabase.auth.verifyOtp({
		email: email.value,
		token: token.value,
		type: 'signup',
	});
}
```
This workaround is OK for me, so this bug can probably be closed, but I will be watching the feature request [supabase/gotrue/issues/905](https://github.com/supabase/gotrue/issues/905)

**Comment by bkan36** [comment]
I've encountered the same issue with reset-password workflow 

my verify OTP return **Email link is invalid or has expired every time**.

I thought the problem was from supabase or SMTP and I fixed it by calling my function at the mounted step of my component
I use nuxt 3 (3.11.2)

here's my code

  `    onMounted(async () => {
        await getUserFromToken()
    })

    async function getUserFromToken() {
        try {
            const tokenHash = route.query.th
            const type = route.query.type

            if (tokenHash && type) {
                const { data, error } = await supabase.auth.verifyOtp({
                    token_hash: tokenHash,
                    type,

                })
                if (error) throw error
                console.log(data)

                email.value = data.user.email

                console.log(email.value)
            }

        } catch (error) {
            console.error(error)
        }
    }`

**Comment by JMarchandev** [comment]
Hi @bkan36 i use same way to verify the TokenHash.
Do you know how to manage lifetime of TokenHash ?
Some auth mail received same day don't match with the verify.

I was looking for on config.toml ?

**Comment by bkan36** [comment]
> Hi @bkan36 i use same way to verify the TokenHash. Do you know how to manage lifetime of TokenHash ? Some auth mail received same day don't match with the verify.
> 
> I was looking for on config.toml ?

I don't know if we can modify the TokanHash. 
The problem with all auth emails is that the token can be verify only once after that it will expire so I thought about a trick to store locally in a cookie the user's id the first time I verify it, then every time the page is reloaded you get the user's id from the cookie so you can extend the lifetime via the expiration of the cookie.

**Comment by JMarchandev** [comment]
@bkan36 Thanks for your answer.
I finally found solution about.
I finally used the otp expiration cause token hash user the verifyOtp method from sdk.

It work for me :)

**Comment by JMarchandev** [comment]
Loccaly in `config.toml`

```
# Number of seconds before the email OTP expires (defaults to 1 hour).
otp_expiry = 86400
```

And in the remote studio: Authentication > Providers > Email
![image](https://github.com/user-attachments/assets/52517797-63ac-4bc2-91ff-7b17623c9827)
